File: models/IADBAirport.m
//
//  Airport.m
//  descent
//
//  Created by Christopher Hobbs on 1/26/14.
//  Copyright (c) 2014 Toonsy Net. All rights reserved.
//

#import <CoreLocation/CoreLocation.h>

#import "IADBAirport.h"
#import "IADBPersistence.h"
#import "IADBModel.h"
#import "IADBFrequency.h"
#import "IADBRunway.h"
#import "IADBConstants.h"

@implementation IADBAirport

@dynamic airportId;
@dynamic code;
@dynamic municipality;


@synthesize frequencies = _frequencies;
@synthesize runways = _runways;

+(NSArray *) types {
    return @[
             AIRPORT_TYPE_LARGE,
             AIRPORT_TYPE_MEDIUM,
             AIRPORT_TYPE_SMALL,
             AIRPORT_TYPE_SEAPLANE,
             AIRPORT_TYPE_HELIPORT,
             AIRPORT_TYPE_BALLOONPORT,
             AIRPORT_TYPE_CLOSED
             ];
}

+(IADBAirport *) findByAirportId:(int32_t) airportId {
    
    NSManagedObjectContext *context = [IADBModel managedObjectContext];
    NSEntityDescription *entityDescription = [NSEntityDescription entityForName:@"IADBAirport" inManagedObjectContext:context];
    NSFetchRequest *request = [[NSFetchRequest alloc] init];
    [request setEntity:entityDescription];
    
    // Set example predicate and sort orderings...
    NSPredicate *predicate = [NSPredicate predicateWithFormat:
                              @"(airportId = %d)",airportId];
    [request setPredicate:predicate];
    
    NSError *error;
    NSArray *airports = [context executeFetchRequest:request error:&error];
    
    NSAssert3(airports, @"Unhandled error removing file in %s at line %d: %@", __FUNCTION__, __LINE__, [error localizedDescription]);
    
    if (!airports || airports.count == 0) { return nil; }
    if (airports.count > 1 ) { NSLog(@"WARNING: more than 1 %@ found with id %d",[self description],airportId); }
    
    return airports[0];
}


// returns locations that begin with identifier with K prepended or the leading K removed
// or that begin with code
+(IADBCenteredArray *) findAllByIdentifierOrCode:(NSString *) identifier withTypes:(NSArray *) types {
    return [self findAllByIdentifierOrCode:identifier orMunicipality:nil withTypes:types];
}


+(IADBCenteredArray *) findAllByIdentifierOrCodeOrMunicipality:(NSString *) identifier withTypes:(NSArray *) types {
    return [self findAllByIdentifierOrCode:identifier orMunicipality:identifier withTypes:types];
}

+(IADBCenteredArray *) findAllByIdentifierOrCode:(NSString *) identifier orMunicipality:(NSString *) municipality withTypes:(NSArray *) types {
    if(!identifier || identifier.length == 0) { return [[IADBCenteredArray alloc] init]; }
    
    NSString *identifierB;
    if( identifier.length >= 2 && [[identifier uppercaseString] hasPrefix:@"K"] ) {
        //allow "KCVH" to find CVH
        identifierB = [identifier substringFromIndex:1]; // strip off K at beginning
    } else {
        identifierB = [NSString stringWithFormat:@"K%@",identifier]; // add K to beginning
    }
    
    return [self findAllByIdentifiers:@[identifier, identifierB] orCode:identifier orMunicipality:municipality withTypes:types];
}

+(IADBCenteredArray *) findAllByIdentifiers:(NSArray *) identifiers orCode:(NSString *) code orMunicipality:(NSString *) municipality withTypes:(NSArray *) types {
    NSMutableArray *arguments = [[NSMutableArray alloc] initWithCapacity:3];
    NSMutableArray *predicates = [[NSMutableArray alloc] initWithCapacity:3];
    
    if( !identifiers ) { identifiers = @[]; }
    for ( NSString *identifier in identifiers ) {
        if ( identifier.length > 0) {
            [predicates addObject: @"(identifier BEGINSWITH[c] %@)"];
            [arguments addObject: identifier];
        }
    }
    
    if ( code && code.length > 0) {
        [predicates addObject: @"(code BEGINSWITH[c] %@)"];
        [arguments addObject: code];
    }
    
    if ( municipality && municipality.length > 0) {
        [predicates addObject: @"(municipality BEGINSWITH[c] %@)"];
        [arguments addObject: municipality];
    }
    
    if( predicates.count == 0) {
        // no inputs results in no outputs
        return [[IADBCenteredArray alloc] init];
    }
    
    NSString *predicateString = [predicates componentsJoinedByString:@" or "];
    
    if (types) {
        predicateString = [NSString stringWithFormat:@"(%@) AND (%@)",predicateString,[self predicateTypes:types]];
    }
    NSPredicate *predicate = [NSPredicate predicateWithFormat:predicateString argumentArray:arguments];
    return [self findAllByPredicate:predicate];
    
}

-(NSArray *) frequencies {
    if( !_frequencies ) {
        _frequencies = [IADBFrequency findAllByAirportId:self.airportId];
    }
    return _frequencies;
}

-(NSArray *) runways {
    if( !_runways ) {
        _runways = [IADBRunway findAllByAirportId:self.airportId];
        for (IADBRunway *runway in _runways) {
            runway.airport = self;
        }
    }
    return _runways;
}

-(BOOL) hasRunways {
    return self.runways && self.runways.count > 0;
}

-(NSInteger) longestRunwayFeet {
    NSInteger length = -1;
    for( IADBRunway *runway in self.runways ) {
        if (runway.closed) {
            continue;
        }
        length = MAX(length, runway.lengthFeet);
    }
    return length;
}

-(IADBRunway *) longestRunway {
    IADBRunway *max = nil;
    for( IADBRunway *runway in self.runways ) {
        if (!max || (max.lengthFeet < runway.lengthFeet) || (max.lengthFeet == runway.lengthFeet && max.widthFeet < runway.widthFeet)) {
            if (!runway.closed) {
                max = runway;
            }
        }
    }
    return max;
}

-(BOOL) hasHardRunway {
    for( IADBRunway *runway in self.runways ) {
        if ([runway isHard] && !runway.closed) {
            return YES;
        }
    }
    return NO;
}

-(IADBFrequency *) frequencyForName:(NSString *) name {
    for (IADBFrequency *f in [self frequencies]) {
        NSString *n = f.name;
        if ([name isEqual:n]) {
            return f;
        }
    }
    return nil;
}

-(NSString *) klessIdentifier {
    if( !self.identifier ) { return @""; }
    if( [[[self.identifier substringToIndex:1] uppercaseString] isEqualToString:@"K"] ) {
        return [self.identifier substringFromIndex:1];
    }
    return self.identifier;
}

-(NSString *) title {
    return [NSString stringWithFormat:@"%@: %@",self.identifier,self.name];
}

-(NSString *) description {
    return [NSString stringWithFormat:@"%@ %@ %lf %lf %@ <%@>", self.identifier, self.name, self.latitude, self.longitude, self.elevationFeet, self.type];
}

-(NSDictionary *) asDictionary {
    return @{@"identifier": self.identifier,
             @"name": self.name,
             @"type": self.type,
             @"latitude": [NSNumber numberWithDouble:self.latitude],
             @"longitude": [NSNumber numberWithDouble:self.latitude],
             @"elevationFeet": self.elevationFeet
             };
}

@end

File: models/IADBAirport.swift
//
//  IADBAirport.swift
//  iAirportsDB
//
//  Created by Christopher Hobbs on 8/21/16.
//  Copyright © 2016 Toonsy Net. All rights reserved.
//

import Foundation
import CoreData

@objc(IADBAirport)
class IADBAirport: IADBLocationElevation {
    //identifier: String (defined in IADBLocation) // identifier or gps code e.g. RKSI, use this for pilot navigation
    @NSManaged var airportId: Int32
    @NSManaged var code: String // iata code e.g. ICN, this is what is printed on boarding passes
    @NSManaged var municipality: String // e.g. Seoul
    
    enum AirportType: String {
        case Large = "large_airport", Medium = "medium_airport", Small = "small_airport",
        Seaplane = "seaplane_base", Heliport = "heliport", Balloonport = "balloonport", Closed = "closed"
        
        static let allValues = [Large, Medium, Small, Seaplane, Heliport, Balloonport, Closed]
    }
    
    class func findByAirportId(airportId: Int32) -> IADBAirport? {
        let context = IADBModel.managedObjectContext()
        let entityDescription = NSEntityDescription.entityForName("IADBAirport", inManagedObjectContext: context)
        let request = NSFetchRequest()
        request.entity = entityDescription!
        // Set example predicate and sort orderings...
        let predicate = NSPredicate(format: "(airportId = %d)", airportId)
        request.predicate = predicate
        
        do {
            let array = try context.executeFetchRequest(request)
            if array.isEmpty {
                return nil
            }
            if let airports = array as? [IADBAirport] {
                if airports.count > 1 {
                    print("WARNING: more than 1 \(self.description) found with id \(airportId)")
                }
                return airports[0]
            } else {
                print("Fetch contained an invalid type \(array)")
            }
        } catch let error as NSError {
            // failure
            print("Fetch failed: \(error.localizedDescription)")
        }
        return nil
    }
    
    // returns locations that begin with identifier with K prepended or the leading K removed
    // or that begin with code
    
    class func findAllByIdentifierOrCode(identifier: String, withTypes types: [String]?) -> IADBCenteredArray {
        return self.findAllByIdentifierOrCode(identifier, orMunicipality: nil, withTypes: types)
    }
    
    class func findAllByIdentifierOrCodeOrMunicipality(identifier: String, withTypes types: [String]?) -> IADBCenteredArray {
        return self.findAllByIdentifierOrCode(identifier, orMunicipality: identifier, withTypes: types)
    }
    
    class func findAllByIdentifierOrCode(identifier: String, orMunicipality municipality: String?, withTypes types: [String]?) -> IADBCenteredArray {
        if identifier.isEmpty {
            return IADBCenteredArray()
        }
        var identifierB = ""
        if identifier.characters.count >= 2 && identifier.uppercaseString.hasPrefix("K") {
            //allow "KCVH" to find CVH
            identifierB = identifier.substringFromIndex(identifier.startIndex.advancedBy(1))
            // strip off K at beginning
        }
        else {
            identifierB = "K\(identifier)"
            // add K to beginning
        }
        return self.findAllByIdentifiers([identifier, identifierB], orCode: identifier, orMunicipality: municipality, withTypes: types)
    }
    // similar to some code in IADBLocation finders
    
    class func findAllByIdentifiers(identifiers: [String], orCode code: String?, orMunicipality municipality: String?, withTypes types: [String]?) -> IADBCenteredArray {
        var arguments = [String]()
        var predicates = [String]()
        for identifier: String in identifiers {
            if !identifier.isEmpty {
                predicates.append("(identifier BEGINSWITH[c] %@)")
                arguments.append(identifier)
            }
        }
        if let _code = code where _code.characters.count > 0 {
            predicates.append("(code BEGINSWITH[c] %@)")
            arguments.append(_code)
        }
        if let _municipality = municipality where _municipality.characters.count > 0 {
            predicates.append("(municipality BEGINSWITH[c] %@)")
            arguments.append(_municipality)
        }
        if predicates.count == 0 {
            // no inputs results in no outputs
            return IADBCenteredArray()
        }
        var predicateString = predicates.joinWithSeparator(" or ")
        predicateString = "(\(predicateString)) AND \(self.predicateTypes(types))"
        let predicate = NSPredicate(format: predicateString, argumentArray: arguments)
        return self.findAllByPredicate(predicate)
    }
    
    lazy var frequencies:[IADBFrequency] = {
        if let models = IADBFrequency.findAllByAirportId(self.airportId) as? [IADBFrequency] {
            return models
        } else {
            print("!!! Invalid frequency type for airport")
        }
        return []
    }()
    
    
    lazy var runways:[IADBRunway] = {
        if let models = IADBRunway.findAllByAirportId(self.airportId) as? [IADBRunway] {
            return models
        } else {
            print("!!! Invalid runway type for airport")
        }
        return []
    }()
    
    func hasRunways() -> Bool {
        return self.runways.count > 0
    }
    
    func longestRunwayFeet() -> Int16 {
        var length:Int16 = -1
        for runway: IADBRunway in self.runways {
            if !runway.closed {
                length = max(length, runway.lengthFeet)
            }
        }
        return length
    }
    
    func longestRunway() -> IADBRunway? {
        var maxRunway: IADBRunway? = nil
        for runway: IADBRunway in self.runways {
            if let max = maxRunway {
                if (max.lengthFeet < runway.lengthFeet) || (max.lengthFeet == runway.lengthFeet && max.widthFeet < runway.widthFeet) {
                    if !runway.closed {
                        maxRunway = runway
                    }
                }
            } else {
                maxRunway = runway
            }
        }
        return maxRunway
    }
    
    func hasHardRunway() -> Bool {
        for runway: IADBRunway in self.runways {
            if runway.isHard() && !runway.closed {
                return true
            }
        }
        return false
    }
    
    func frequencyForName(name: String) -> IADBFrequency? {
        for f: IADBFrequency in self.frequencies {
            if name.isEqual(f.name) {
                return f
            }
        }
        return nil
    }
    
    func klessIdentifier() -> String {
        if (self.identifier.substringToIndex(self.identifier.startIndex.advancedBy(1)).uppercaseString == "K") {
            return self.identifier.substringFromIndex(self.identifier.startIndex.advancedBy(1))
        }
        return self.identifier
    }
    
    func title() -> String {
        return "\(self.identifier): \(self.name)"
    }
    
    override var description: String {
        return "<\(self.identifier) (\(self.code)) \(self.name) \(self.latitude) \(self.longitude) \(self.elevationFeet) <\(self.type)>"
    }
    
    func asDictionary() -> [NSObject : AnyObject] {
        return ["identifier": self.identifier, "name": self.name, "type": self.type, "latitude": Int(self.latitude), "longitude": Int(self.latitude), "elevationFeet": self.elevationForLocation()]
    }
    
    override func setCsvValues( values: [String: String] ) {
        //"id","ident","type","name","latitude_deg","longitude_deg","elevation_ft","continent","iso_country","iso_region","municipality","scheduled_service","gps_code","iata_code","local_code","home_link","wikipedia_link","keywords"
        //print(values)
        
        self.airportId = Int32(values["id"]!)!
        self.code = values["iata_code"] ?? ""
        let elevationString = values["elevation_ft"] ?? ""
        self.identifier = (values["gps_code"] ?? "").isEmpty ? values["ident"] ?? "" : values["gps_code"] ?? ""
        self.latitude = Double(values["latitude_deg"]!)!
        self.longitude = Double(values["longitude_deg"]!)!
        self.municipality = values["municipality"] ?? ""
        self.name = values["name"] ?? ""
        self.type = values["type"] ?? ""
        self.elevationFeet = elevationString.isEmpty ? nil : Int( elevationString )
    }

}
